iceoryx_posh/doc/chunk_header.md

```diff
@@ -0,0 +1,218 @@
+
+```
+headerEndAddress = addressof(chunkHeader) + sizeof(chunkHeader);
+payloadAddress = align(headerEndAddress, payloadAlignment);
+payloadOffset = payloadAddress - addressof(chunkHeader);
+```
+
+3. Custom header is used
+
+```
+headerEndAddress = addressof(chunkHeader) + sizeof(chunkHeader) + sizeof(customHeader);
+potentialBackOffsetAddress = align(headerEndAddress, alignof(payloadOffset));
+potentialPayloadAddress = potentialBackOffsetAddress + sizeof(payloadOffset);
+payloadAddress = align(potentialPayloadAddress, payloadAlignment);
+payloadOffset = payloadAddress - addressof(chunkHeader);
+```
+
+### Required Chunk Size Calculation
+
+In order to fit the custom header and the payload into he chunk, a worst case calculation has to be done. We can assume that a chunk is aligned to 32 bytes, which is also the alignment of the `ChunkHeader`.
+
+1. No custom header and payload alignment doesn't exceed the `ChunkHeader` alignment
+
+```
+chunkSize = sizeof(chunkHeader) + payloadSize;
+```
+
+2. No custom header and payload alignment exceeds the `ChunkHeader` alignment
+
+Worst case scenario is when a part of the `ChunkHeader` crosses the payload alignment boundary, so that the payload must be aligned to the next boundary.
+```
+sizeLeftOfThePayloandAlignmentBoundary = sizeof(chunkHeader) - alignof(chunkHeader);
+chunkSize = sizeLeftOfThePayloandAlignmentBoundary + payloadAlignment + payloadSize;
+```
+
+3. Custom header is used
+
+Similar to case 2, but in this case it is the `back-offset` which might cross the payload alignment boundary.
+```
+headerSize = sizeof(chunkHeader) + sizeof(customHeader)
+sizeLeftOfThePayloandAlignmentBoundary = align(headerSize, alignof(payloadOffset));
+maxAlignment = max(alignof(payloadOffset), payloadAlignment);
+chunkSize = sizeLeftOfThePayloandAlignmentBoundary + maxAlignment + payloadSize;
+```
+
+### Accessing Chunk Header Extension
+
+The `ChunkHeader` has a template method to get custom header. There is a risk to use this wrong by accident, but there is currently no better solution for this.
+
+Since the custom header is always adjacent to the `ChunkHeader`, the formula to get the custom header is:
+```
+customHeader = addressOf(chunkHeader) + sizeof(chunkHeader);
+```
+
+### ChunkHeader Methods
+
+- `void* payload()` returns a pointer to the payload
+- `template <typename T> T* customHeader()` returns a pointer to the custom header
+- `static ChunkHeader* fromPayload(const void* const payload)` returns a pointer to the `ChunkHeader` associated to the payload
+
+### Integration Into Publisher/Subscriber API
+
+The `Publisher` has additional template parameters for the custom header and payload alignment. Alternatively this could also be done with the `allocate` method, but that increases the risk of using it wrong by accident.
```

Comment:
For me this is fixed for a specific publisher. So no need to have it in the allocate call

Reply:
@budrus I would disagree here.
Isn't the idea of the untyped publisher to acquire memory for any arbitrary type and the type and amount of memory can change with every call? With this also the alignment can change with every call!
For me it seems quite natural to have then a second argument in allocate the define the alignment.
`untypedPublisher.allocate(memSize, alignment)`

In the case of the typed publisher I totally agree with you but here we just call `loan`.

Reply:
I think elfenpiff is right. After thinking a bit about this, it might be helpful to have this in allocate. Maybe we have to write down some use cases for the untyped publisher.

Reply:
Puuh, You really see a use case where one publisher is used with different alignments for different samples? Quite academic IMHO. If you need a special alignment that is > 32 byte, I would assume that every sample would have the same alignment as there is always the same processing for each sample and this has a specific alignment requirement

Reply:
Let's say you would like to send samples which are containing data for vector operations implemented with AVX512 or SSE2/3/4. Then you need an alignment of 64 bytes. And currently we explicitly allow the user to send a bool as next sample where an alignment of 1 would suffice.

I think we should go all in in this case and support really the stuff the untyped publisher promises the expert user.

Reply:
Like elfenpiff said, we can already specify the payload size when a sample is loaned, therefore it makes sense to also be able to specify the alignment. I would specify the custom header with the publisher/subscriber, though.

Reply:
The untyped subscriber is very likely used when iceoryx is integrated in a bigger framework like eCAL. At least I would go for a default as for most of them it will be don't care

Reply:
yes, default alignment would be 8 bytes

Reply:
OK, go for it